triplet/main.py

```diff
@@ -0,0 +1,487 @@
+            """
+            if transform is not None:
+                img = self.transform(img.numpy())
+            return img
+
+        if not self.train:
+            m = self.matches[index]
+            img1, img2 = transform(self.data[m[0]]), transform(self.data[m[1]])
+            return img1, img2, m[2]
+
+        t = self.triplets[index]
+        a, p, n = self.data[t[0]], self.data[t[1]], self.data[t[2]]
+
+        # transform images if required
+        img_a, img_p, img_n = transform(a), transform(p), transform(n)
+        return img_a, img_p, img_n
+
+    def __len__(self):
+        if self.train:
+            return self.triplets.size(0)
+        else:
+            return self.matches.size(0)
+
+
+class TNet(nn.Module):
+    """TFeat model definition
+    """
+    def __init__(self):
+        super(TNet, self).__init__()
+        self.features = nn.Sequential(
+            nn.Conv2d(1, 32, kernel_size=7),
+            nn.Tanh(),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            nn.Conv2d(32, 64, kernel_size=6),
+            nn.Tanh()
+        )
+        self.classifier = nn.Sequential(
+            nn.Linear(64 * 8 * 8, 128),
+            nn.Tanh()
+        )
+
+    def forward(self, x):
+        x = self.features(x)
+        x = x.view(x.size(0), -1)
+        x = self.classifier(x)
+        return x
+
+
+def weights_init(m):
+    # From https://github.com/alykhantejani/nninit/blob/master/nninit.py#L66
+    def _calculate_fan_in_and_fan_out(tensor):
+        if tensor.ndimension() < 2:
+            raise ValueError(
+               "fan in and fan out can not be computed for tensor of size ",
+               tensor.size())
+
+        if tensor.ndimension() == 2:  # Linear
+            fan_in = tensor.size(1)
+            fan_out = tensor.size(0)
+        else:
+            num_input_fmaps = tensor.size(1)
+            num_output_fmaps = tensor.size(0)
+            receptive_field_size = np.prod(tensor.numpy().shape[2:])
+            fan_in = num_input_fmaps * receptive_field_size
+            fan_out = num_output_fmaps * receptive_field_size
+
+        return fan_in, fan_out
+
+    def xavier_uniform(data, gain=1):
+        fan_in, fan_out = _calculate_fan_in_and_fan_out(data)
+        std = gain * np.sqrt(2.0 / (fan_in + fan_out))
+        a = np.sqrt(3.0) * std
+        data.uniform_(-a, a)
+
+    def constant(tensor, val):
+        tensor.fill_(val)
+
+    if isinstance(m, nn.Conv2d):
+        xavier_uniform(m.weight.data)
+        constant(m.bias.data, 0.1)
+
+
+class PairwiseDistance(Function):
+    def __init__(self, p):
+        super(PairwiseDistance, self).__init__()
+        self.norm = p
+
+    def forward(self, x1, x2):
+        assert x1.size() == x2.size()
+        eps = 1e-4 / x1.size(1)
+        diff = torch.abs(x1 - x2)
+        out = torch.pow(diff, self.norm).sum(dim=1)
+        return torch.pow(out + eps, 1. / self.norm)
+
+
+class TripletMarginLoss(Function):
```

Comment:
You should subclass a `Function` only if you want to extend autograd, which is not what you're doing in here (you haven't defined `backward`, so it's not a valid one). If you want to compose some autograd ops either use a regular Python function declared with `def`, or subclass from `object` or `nn.Module`

Same goes for `TripletMarginLossSwap` below.